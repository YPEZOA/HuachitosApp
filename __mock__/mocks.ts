import { AnimalProps } from '../src/utils/models'

export const animals: AnimalProps[] = [
  {
    id: 1,
    nombre: 'First Animal',
    tipo: 'Gato',
    color: 'Cafe',
    edad: '9 Meses',
    estado: 'Adopcion',
    genero: 'Hembra',
    desc_fisica: 'Lorem Ipsum Lorem Ipsum',
    desc_personalidad: 'Lorem Ipsum in person lorem Ipsum',
    desc_adicional: 'Lorem ipsum dolor sit am',
    esterilizado: 1,
    vacunas: 1,
    imagen: '',
    equipo: '',
    region: 'RM',
    comuna: 'Puente Alto',
    url: '',
  },
  {
    id: 2,
    nombre: 'Second Animal',
    tipo: 'Perro',
    color: 'Negro',
    edad: '5 Meses',
    estado: 'Adopcion',
    genero: 'Macho',
    desc_fisica: 'Lorem Ipsum Lorem Ipsum',
    desc_personalidad: 'Lorem Ipsum in person lorem Ipsum',
    desc_adicional: 'Lorem ipsum dolor sit am',
    esterilizado: 1,
    vacunas: 1,
    imagen: '',
    equipo: '',
    region: 'RM',
    comuna: 'Puente Alto',
    url: '',
  },
  {
    id: 3,
    nombre: 'Third Animal',
    tipo: 'Conejo',
    color: 'Plateado',
    edad: '4 Meses',
    estado: 'Adopcion',
    genero: 'Hembra',
    desc_fisica: 'Lorem Ipsum Lorem Ipsum',
    desc_personalidad: 'Lorem Ipsum in person lorem Ipsum',
    desc_adicional: 'Lorem ipsum dolor sit am',
    esterilizado: 1,
    vacunas: 1,
    imagen: '',
    equipo: '',
    region: 'RM',
    comuna: 'Santiago',
    url: '',
  },
]
