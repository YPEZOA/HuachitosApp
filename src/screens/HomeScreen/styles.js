import { StyleSheet } from 'react-native'
import colors from '../../UI/colors'

export const homeScreen = StyleSheet.create({
  containerMain: {
    flex: 1,
    paddingTop: 50,
    paddingHorizontal: 10,
  },
  scrollContainer: {
    marginTop: 30,
  },
  inputContainer: {
    marginTop: 20,
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
    padding: 5,
    backgroundColor: '#dcdcdc',
    borderRadius: 100,
  },
  inputIcon: {
    padding: 10,
    borderRadius: 100,
    backgroundColor: colors.white,
  },
  headerLogoContent: {
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',
  },
  headerText: {
    fontSize: 28,
    fontFamily: 'Quicksand-Bold',
    color: colors.primary,
  },
  title: {
    marginTop: 20,
    fontSize: 25,
    fontWeight: 'bold',
    color: '#5e5e5e',
  },
})
