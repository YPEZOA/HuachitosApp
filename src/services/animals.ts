const URI = 'https://huachitos.cl/api'

export const getAllAnimals = () => {
  try {
    const response = fetch(`${URI}/animales`).then(res => res.json())
    return response
  } catch (error) {
    console.error(error)
  }
}

export const getAnimalsByType = (type: string) => {
  try {
    const response = fetch(`${URI}/animales/tipo/${type}`).then(res =>
      res.json()
    )
    return response
  } catch (error) {
    console.error(error)
  }
}
