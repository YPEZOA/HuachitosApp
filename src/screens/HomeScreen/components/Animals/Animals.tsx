import React, { useContext } from 'react'
import { ActivityIndicator, Text, View } from 'react-native'
import { HomeContext } from '../../context'
import Animal from '../Animal/Animal'
import MasonryList from '@react-native-seoul/masonry-list'
import Animated, { FadeInDown } from 'react-native-reanimated'
import { animalsStyles as St } from '../Animals/styles'
import colors from '../../../../UI/colors'

const Animals = () => {
  const { states } = useContext(HomeContext)

  if (!states.allAnimals.length && !states.fetching)
    return (
      <Animated.View
        entering={FadeInDown.springify().duration(1000).damping(20)}
        style={St.loaderContainer}
      >
        <Text style={St.notRestultText}>No se encontraron resultados</Text>
      </Animated.View>
    )

  return (
    <View style={St.container}>
      {states.fetching ? (
        <View style={St.loaderContainer}>
          <ActivityIndicator color={colors.secondary} size="large" />
        </View>
      ) : (
        <View style={{ flex: 1 }}>
          <MasonryList
            ListHeaderComponent={
              <Text style={St.titleList}>Esperan por ti</Text>
            }
            refreshControl={false}
            keyExtractor={item => item.id}
            showsVerticalScrollIndicator={false}
            data={states.allAnimals}
            renderItem={({ item, i }: any) => <Animal data={item} index={i} />}
            numColumns={2}
            onEndReachedThreshold={0.1}
            style={{ gap: 10 }}
          />
        </View>
      )}
    </View>
  )
}

export default Animals
