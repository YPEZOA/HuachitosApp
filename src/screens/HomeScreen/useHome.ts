import { useEffect, useState } from 'react'
import { getAllAnimals, getAnimalsByFilters, getAnimalsByType } from '../../services/animals'
import { isFetching } from '../../utils/helpers/loader'

const initialState = {
  state: '',
  region: { id: null, region: '', comunas: [] },
  comuna: null,
}

const useHome = () => {
  const [allAnimals, setAllAnimals] = useState([])
  const [animalTypeSelected, setAnimalTypeSelected] = useState('')
  const [fetching, setFetching] = useState(false)
  const [showFiltersModal, setShowFiltersModal] = useState(false)
  const [filterParameters, setFilterParameters] = useState(initialState)
  const [animalsFiltered, setAnimalsFiltered] = useState([])
  const [filterSubmited, setFilterSubmited] = useState(false)
  const [fetchWithoutData, setFetchWithoutData] = useState(false)

  const fetchWrap = isFetching(setFetching)

  useEffect(() => {
    async function getAnimals() {
      setFetching(true)
      const animals = await getAllAnimals()
      dataIsVisible(animals)
      setAllAnimals(animals.data)
      setFetching(false)
    }
    getAnimals()
  }, [])

  const dataIsVisible = ({ data }: any) => {
    return !data?.length ? setFetchWithoutData(true) : setFetchWithoutData(false)
  }

  const resetFilter = () => {
    setFilterSubmited(false)
    setFilterParameters(initialState)
    setAnimalsFiltered([])
    setFilterSubmited(false)
  }

  return {
    states: {
      fetchWithoutData,
      filterSubmited,
      allAnimals,
      animalTypeSelected,
      fetching,
      showFiltersModal,
      filterParameters,
      animalsFiltered,
      get getAvailableCommunes() {
        return filterParameters.region.comunas
      },
      get anyFilterSelected() {
        return filterParameters !== initialState
      },
      get filterResultLength() {
        return animalsFiltered?.length > 0
      },
    },
    setters: {
      setFilterSubmited,
      setAnimalTypeSelected,
      setShowFiltersModal,
      setFilterParameters,
      setAnimalsFiltered,
    },
    actions: {
      resetFilter,
      filterAnimalsByType: fetchWrap(async (type: string) => {
        if (!type.length) return
        const animalsFilter = await getAnimalsByType(type)
        dataIsVisible(animalsFilter)
        setAnimalsFiltered(animalsFilter.data)
      }),
      getAnimalsByFiltered: fetchWrap(async () => {
        const filterPayload = {
          ...filterParameters,
          region: filterParameters.region.id,
        }
        const { comuna, state, region } = filterPayload

        if (filterPayload.comuna !== null) {
          const animalsByComuna = await getAnimalsByFilters(
            `comuna/${comuna}${state.length ? `/estado/${state}` : ''}`
          )
          dataIsVisible(animalsByComuna)
          setAnimalsFiltered(animalsByComuna.data)
          setAnimalTypeSelected('')
          setFilterParameters(initialState)
        } else {
          const regionSelected = filterPayload.region !== null
          const stateSelected = filterPayload.state !== ''
          const animalsFilteredBySelection = await getAnimalsByFilters(
            `${
              regionSelected && stateSelected
                ? `region/${region}/estado/${state}`
                : regionSelected
                ? `region/${region}`
                : `estado/${state}`
            }`
          )
          setAnimalTypeSelected('')
          setAnimalsFiltered(animalsFilteredBySelection.data)
          dataIsVisible(animalsFilteredBySelection)
          setFilterParameters(initialState)
        }
      }),
    },
  }
}

export default useHome
