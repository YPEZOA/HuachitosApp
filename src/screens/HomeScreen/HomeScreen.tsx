import React, { useEffect } from 'react'
import { useNavigation } from '@react-navigation/native'
import { Text, View } from 'react-native'
import { homeScreen as St } from './styles'
import { getAllAnimals } from '../../utils/helpers/http'
import Animals from './components/Animals/Animals'
import Icon from 'react-native-vector-icons/FontAwesome'
import colors from '../../UI/colors'

const HomeScreen = () => {
  useEffect(() => {
    getAllAnimals()
  }, [])

  const navigation = useNavigation()

  return (
    <View style={St.containerMain}>
      <View style={St.headerLogoContent}>
        <Text style={St.headerText}>
          Huachit
          <Icon name="paw" size={20} />s
        </Text>
        <Text style={{ color: colors.primary, fontSize: 10 }}>
          Encuentra y da familia
        </Text>
      </View>
      <Animals />
    </View>
  )
}

export default HomeScreen
