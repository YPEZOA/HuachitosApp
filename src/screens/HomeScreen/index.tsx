import React from 'react'
import { TouchableOpacity, View, StatusBar, Pressable, Text } from 'react-native'
import { homeScreen as St } from './styles'
import { HomeContext } from './context'
import useHome from './useHome'
import Animals from './components/Animals/Animals'
import IconF from 'react-native-vector-icons/Ionicons'
import Filters from './components/Filter/Filters'
import FiltersModal from './components/FiltersModal/FiltersModal'
import LightLogoText from '../../UI/Icons/LogoWithTagLight'
import FilterResult from './components/FilterResult/FilterResult'
import { useNavigation } from '@react-navigation/native'
import colors from '../../UI/colors'

const HomeScreen = () => {
  const { states, actions, setters } = useHome()
  const navigation = useNavigation()

  return (
    <HomeContext.Provider value={{ states, actions, setters }}>
      <StatusBar backgroundColor={colors.primary} />
      {/* <Pressable onPress={() =>  */}
      {/*   <Text>FAVORITES</Text> */}
      {/* </Pressable> */}
      <View style={St.containerMain}>
        {/* Header */}
        <View style={St.headerContainer}>
          <LightLogoText width={220} height={70} />
          <View style={St.headerButtonsContainer}>
            <TouchableOpacity
              onPress={() => navigation.navigate('AnimalsLikes')}
              style={St.closeIcon}
            >
              <IconF name="heart" size={30} color="#ee6352" />
            </TouchableOpacity>
            <TouchableOpacity
              onPress={() => setters.setShowFiltersModal(true)}
              style={St.closeIcon}
            >
              <IconF name="filter" size={30} color={colors.secondary} />
            </TouchableOpacity>
          </View>
        </View>
        {/* Animals list */}
        <View style={St.containerContent}>
          <Filters />
          {states.filterSubmited ? <FilterResult /> : <Animals />}
        </View>
      </View>
      {/* Filters Modal */}
      <FiltersModal />
    </HomeContext.Provider>
  )
}

export default HomeScreen
