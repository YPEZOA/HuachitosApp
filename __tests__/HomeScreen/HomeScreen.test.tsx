import React from 'react'
import HomeScreen from '../../src/screens/HomeScreen'
import { fireEvent, render, renderHook, waitFor } from '@testing-library/react-native'
import useHome from '../../src/screens/HomeScreen/useHome'
import { HomeContext } from '../../src/screens/HomeScreen/context'
import Animals from '../../src/screens/HomeScreen/components/Animals/Animals'
import * as api from '../../src/services/animals'
import { animals } from '../../__mock__/mocks'

const mockedNavigation = jest.fn()
jest.mock('@react-navigation/native', () => {
  return {
    useNavigation: () => ({
      navigate: mockedNavigation.mockImplementation(() => {}),
    }),
  }
})

describe('<HomeScreen/> test', () => {
  afterEach(() => jest.clearAllMocks())
  test('Should by render correctly', () => {
    const container = render(<HomeScreen />)
    expect(container).toBeTruthy()
  })

  test('Should be show filter and animals liked buttons', () => {
    const { getByLabelText } = render(<HomeScreen />)
    const likesButton = getByLabelText('Ir a favoritos')
    const openFiltersButton = getByLabelText('Abrir filtros')

    fireEvent.press(likesButton)
    fireEvent.press(openFiltersButton)
    expect(likesButton).toBeDefined()
    expect(openFiltersButton).toBeDefined()
  })

  test('Should be render list buttons of animals type filter', () => {
    const { getAllByTestId } = render(<HomeScreen />)
    const button = getAllByTestId('animalType')
    expect(button).toBeDefined()
    expect(button.length).toEqual(5)
  })

  test('Should be render <WithoutResult/>', async () => {
    const { result } = renderHook(() => useHome())
    const { states, setters, actions } = result.current

    states.fetchWithoutData = true
    states.fetching = false

    const { getByText } = render(
      <HomeContext.Provider value={{ states, setters, actions }}>
        <Animals />
      </HomeContext.Provider>
    )
    expect(getByText('No se encontraron resultados')).toBeTruthy()
  })

  test('Should be render animals list', async () => {
    const fetchingSpy = jest
      .spyOn(api, 'getAllAnimals')
      .mockImplementation(() => Promise.resolve({ data: animals }))

    const { getByTestId, getAllByTestId } = render(<HomeScreen />)
    await waitFor(() => expect(fetchingSpy).toHaveBeenCalledTimes(1))

    expect(getByTestId('animals-list')).toBeTruthy()
    expect(getAllByTestId('animal')).toHaveLength(animals.length)
  })

  test('Should navigate to a specific animal when pressing that box', async () => {
    const fetchingSpy = jest
      .spyOn(api, 'getAllAnimals')
      .mockImplementation(() => Promise.resolve({ data: animals }))

    const { getAllByTestId } = render(<HomeScreen />)
    await waitFor(() => expect(fetchingSpy).toHaveBeenCalledTimes(1))

    const animalProfileNavigate = getAllByTestId('animal-profile-navigate')[0]
    fireEvent.press(animalProfileNavigate)
    expect(mockedNavigation).toHaveBeenCalledWith('AnimalProfile', animals[0])
  })
})
