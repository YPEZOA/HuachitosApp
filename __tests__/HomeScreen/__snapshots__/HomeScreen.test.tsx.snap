// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should be to call api when press type of animal 1`] = `[Function]`;
